clients/command.go

```diff
@@ -0,0 +1,302 @@
+	messageBus      messaging.MessageClient
+	queryMessages   chan types.MessageEnvelope
+	queryErrors     chan error
+	commandMessages chan types.MessageEnvelope
+	commandErrors   chan error
+	topics          map[string]string
+	timeout         time.Duration
+}
+
+func NewCommandClient(messageBus messaging.MessageClient, topics map[string]string, timeout time.Duration) (interfaces.CommandClient, error) {
+	client := &CommandClient{
+		messageBus: messageBus,
+		topics:     topics,
+		timeout:    timeout,
+	}
+
+	queryResponseTopic, ok := topics[QueryResponseTopic]
+	if ok {
+		queryMessages := make(chan types.MessageEnvelope, 1)
+		queryErrors := make(chan error)
+		queryTopics := []types.TopicChannel{
+			{
+				Topic:    queryResponseTopic,
+				Messages: queryMessages,
+			},
+		}
+		err := messageBus.Subscribe(queryTopics, queryErrors)
+		if err != nil {
+			return nil, err
+		}
+
+		client.queryMessages = queryMessages
+		client.queryErrors = queryErrors
+	}
+
+	commandResponseTopic, ok := topics[CommandResponseTopic]
+	if ok {
+		commandMessages := make(chan types.MessageEnvelope, 1)
+		commandErrors := make(chan error)
+		commandTopics := []types.TopicChannel{
+			{
+				Topic:    commandResponseTopic,
+				Messages: commandMessages,
+			},
+		}
+		err := messageBus.Subscribe(commandTopics, commandErrors)
+		if err != nil {
+			return nil, err
+		}
+
+		client.commandMessages = commandMessages
+		client.commandErrors = commandErrors
+	}
+
+	return client, nil
+}
+
+func (c *CommandClient) AllDeviceCoreCommands(ctx context.Context, offset int, limit int) (responses.MultiDeviceCoreCommandsResponse, edgexErr.EdgeX) {
+	if c.queryMessages == nil {
+		return responses.MultiDeviceCoreCommandsResponse{}, edgexErr.NewCommonEdgeX(edgexErr.KindServerError, "commandquery request/response topics not provided", nil)
+	}
+
+	queryParams := map[string]string{common.Offset: strconv.Itoa(offset), common.Limit: strconv.Itoa(limit)}
+	requestEnvelope := types.NewMessageEnvelopeForRequest(nil, queryParams)
+	requestTopic := strings.Join([]string{c.topics[QueryRequestTopic], common.All}, "/")
+	err := c.messageBus.Publish(requestEnvelope, requestTopic)
+	if err != nil {
+		return responses.MultiDeviceCoreCommandsResponse{}, edgexErr.NewCommonEdgeXWrapper(err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return responses.MultiDeviceCoreCommandsResponse{}, nil
+		case <-time.After(c.timeout):
+			return responses.MultiDeviceCoreCommandsResponse{}, edgexErr.NewCommonEdgeX(edgexErr.KindServerError, "timed out waiting response", nil)
+		case err = <-c.queryErrors:
+			return responses.MultiDeviceCoreCommandsResponse{}, edgexErr.NewCommonEdgeXWrapper(err)
+		case responseEnvelope := <-c.queryMessages:
+			if responseEnvelope.ErrorCode == 1 {
+				return responses.MultiDeviceCoreCommandsResponse{}, edgexErr.NewCommonEdgeXWrapper(errors.New(string(responseEnvelope.Payload)))
+			}
+			if responseEnvelope.RequestID != requestEnvelope.RequestID {
```

Comment:
always checking the requestID before checking the errorCode, because the error might belong to another command or query request

Reply:
@cloudxxx8 , great catch!